cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramNotificationSubscriberService.java

```diff
@@ -134,6 +142,55 @@
     this.tasks = new LinkedList<>();
     this.metricsCollectionService = metricsCollectionService;
     this.programCompletionNotifiers = Collections.emptySet();
+    this.cConf = cConf;
+    this.txRunner = transactionRunner;
+    this.store = store;
+  }
+
+  @Override
+  protected void doStartUp() throws Exception {
+    super.doStartUp();
+    AtomicBoolean completed = new AtomicBoolean();
+    AtomicReference<AppMetadataStore.Cursor> cursorRef = new AtomicReference<>(AppMetadataStore.Cursor.EMPTY);
+    int limit = cConf.getInt(Constants.RuntimeMonitor.INIT_BATCH_SIZE);
+    long startTs = System.currentTimeMillis();
+
+    try {
+      while (!completed.get()) {
+        TransactionRunners.run(txRunner, context -> {
+          AppMetadataStore store = AppMetadataStore.create(context);
+          completed.set(true);
+          store.scanActiveRuns(cursorRef.get(), limit, (cursor, runRecordDetail) -> {
+            if (runRecordDetail.getStartTs() > startTs) {
```

Comment:
Major:  Looking at the code, it seems run records are stored in descending order of timestamp, meaning scan returns the latest runrecord first.  But the logic here is written with the assumption of run record processed in ascending timestamp. 

Could you run an experiment or using unit test for store to verify the order in which runrecord is returned? If the above is confirmed, remote execution twill runner service has a similar code which may have the same bug.

Reply:
Good catch. We should make sure that the order in which run records are returned has no effect on the correctness of re-trying STARTING pipelines (or similarly creating controllers). Fixed the issue by setting cursorRef and completed flag at the beginning of BiConsumer.

Reply:
Will fix the similar bug in RemoteExecutionTwillRunnerService.java in a separate PR.